src/rulesets/LinearAlgebra/symmetric.jl

```diff
@@ -77,3 +72,201 @@ function _hermitian_back(ΔΩ::LinearAlgebra.AbstractTriangular, uplo)
+        return NO_FIELDS, ∂A
+    end
+    eigen_pullback(ΔF::AbstractZero) = (NO_FIELDS, ΔF)
+    return F, eigen_pullback
+end
+
+# ∂U is overwritten if not an `AbstractZero`
+function eigen_rev!(A::LinearAlgebra.RealHermSymComplexHerm, λ, U, ∂λ, ∂U)
+    ∂λ isa AbstractZero && ∂U isa AbstractZero && return ∂λ + ∂U
+    ∂A = similar(A, eltype(U))
+    tmp = ∂U
+    if ∂U isa AbstractZero
+        mul!(∂A.data, U, real.(∂λ) .* U')
+    else
+        _eigen_norm_phase_rev!(∂U, A, U)
+        ∂K = mul!(∂A.data, U', ∂U)
+        ∂K ./= λ' .- λ
+        ∂K[diagind(∂K)] .= real.(∂λ)
+        mul!(tmp, ∂K, U')
+        mul!(∂A.data, U, tmp)
+        @inbounds _hermitrize!(∂A.data)
+    end
+    return ∂A
+end
+
+_eigen_norm_phase_fwd!(∂V, ::Union{Symmetric{T,S},Hermitian{T,S}}, V) where {T<:Real,S} = ∂V
+function _eigen_norm_phase_fwd!(∂V, A::Hermitian{<:Complex}, V)
+    k = A.uplo === 'U' ? size(A, 1) : 1
+    @inbounds for i in axes(V, 2)
+        v, ∂v = @views V[:, i], ∂V[:, i]
+        vₖ, ∂vₖ = real(v[k]), ∂v[k]
+        ∂v .-= v .* (im * (imag(∂vₖ) / ifelse(iszero(vₖ), one(vₖ), vₖ)))
+    end
+    return ∂V
+end
+
+_eigen_norm_phase_rev!(∂V, ::Union{Symmetric{T,S},Hermitian{T,S}}, V) where {T<:Real,S} = ∂V
+function _eigen_norm_phase_rev!(∂V, A::Hermitian{<:Complex}, V)
+    k = A.uplo === 'U' ? size(A, 1) : 1
+    @inbounds for i in axes(V, 2)
+        v, ∂v = @views V[:, i], ∂V[:, i]
+        vₖ = real(v[k])
+        ∂c = dot(v, ∂v)
+        ∂v[k] -= im * (imag(∂c) / ifelse(iszero(vₖ), one(vₖ), vₖ))
+    end
+    return ∂V
+end
+
+#####
+##### `eigvals!`/`eigvals`
+#####
+
+function frule(
+    (_, ΔA),
+    ::typeof(eigvals!),
+    A::LinearAlgebra.RealHermSymComplexHerm{<:BLAS.BlasReal,<:StridedMatrix};
+    sortby::Union{Function,Nothing}=nothing,
+)
+    ΔA isa AbstractZero && return eigvals!(A; sortby=sortby), ΔA
+    F = eigen!(A; sortby=sortby)
+    λ, U = F.values, F.vectors
+    tmp = ΔA * U
+    # diag(U' * tmp) without computing matrix product
+    ∂λ = similar(λ)
+    @inbounds for i in eachindex(λ)
+        ∂λ[i] = @views real(dot(U[:, i], tmp[:, i]))
+    end
+    return λ, ∂λ
+end
+
+function rrule(
+    ::typeof(eigvals),
+    A::LinearAlgebra.RealHermSymComplexHerm{<:BLAS.BlasReal,<:StridedMatrix};
+    sortby::Union{Function,Nothing}=nothing,
+)
+    F = eigen(A; sortby=sortby)
+    λ = F.values
+    function eigvals_pullback(Δλ)
+        U = F.vectors
+        ∂A = similar(A)
+        mul!(∂A.data, U, real.(Δλ) .* U')
+        return NO_FIELDS, ∂A
+    end
+    eigvals_pullback(Δλ::AbstractZero) = (NO_FIELDS, Δλ)
+    return λ, eigvals_pullback
+end
+
+#####
+##### `svd`
+#####
+
+function rrule(::typeof(svd), A::LinearAlgebra.RealHermSymComplexHerm{<:BLAS.BlasReal,<:StridedMatrix})
+    F = svd(A)
+    function svd_pullback(ΔF::Composite{<:SVD})
+        U, V = F.U, F.V
+        c = _svd_eigvals_sign!(similar(F.S), U, V)
+        λ = F.S .* c
+        ∂λ = ΔF.S isa AbstractZero ? ΔF.S : ΔF.S .* c
```

Comment:
For some reason this is type-unstable:
```julia
julia> using ChainRules, LinearAlgebra

julia> F = svd(randn(4, 4));

julia> ∂F = Composite{typeof(F)}(U = randn(4, 4), V = randn(4, 4)', Vt = randn(4, 4), S = randn(4));

julia> foo(ΔF) = ΔF.U, ΔF.S, ΔF.V, ΔF.Vt;

julia> @inferred foo(∂F)
ERROR: return type Tuple{Array{Float64,2},Array{Float64,1},Adjoint{Float64,Array{Float64,2}},Array{Float64,2}} does not match inferred return type Tuple{Array{Float64,2},Union{Zero, Array{Float64,1}},Union{Zero, Adjoint{Float64,Array{Float64,2}}},Union{Zero, Array{Float64,2}}}
```

Reply:
```julia
julia> @code_warntype getproperty(ChainRulesCore.backing(∂F), :U)
Variables
  #self#::Core.Compiler.Const(getproperty, false)
  x::NamedTuple{(:U, :V, :Vt, :S),Tuple{Array{Float64,2},Adjoint{Float64,Array{Float64,2}},Array{Float64,2},Array{Float64,1}}}
  f::Symbol

Body::Union{Adjoint{Float64,Array{Float64,2}}, Array}
1 ─ %1 = Base.getfield(x, f)::Union{Adjoint{Float64,Array{Float64,2}}, Array}
└──      return %1
```
Looks like this call to `getfield` isn't inferring properly. Possibly something to do with constant-folding or something...

Reply:
Looking at this:
```julia
julia> @code_warntype getproperty(∂F, :U)
Variables
  #self#::Core.Compiler.Const(getproperty, false)
  comp::Composite{SVD{Float64,Float64,Array{Float64,2}},NamedTuple{(:U, :V, :Vt, :S),Tuple{Array{Float64,2},Adjoint{Float64,Array{Float64,2}},Array{Float64,2},Array{Float64,1}}}}
  idx::Symbol

Body::Union{Zero, Adjoint{Float64,Array{Float64,2}}, Array}
1 ─ %1 = (idx ∈ $(Expr(:static_parameter, 2)))::Bool
└──      goto #3 if not %1
2 ─      goto #4
3 ─ %4 = ChainRulesCore.Zero()::Core.Compiler.Const(Zero(), false)
└──      return %4
4 ─ %6 = ChainRulesCore.backing(comp)::NamedTuple{(:U, :V, :Vt, :S),Tuple{Array{Float64,2},Adjoint{Float64,Array{Float64,2}},Array{Float64,2},Array{Float64,1}}}
│   %7 = ChainRulesCore.getproperty(%6, idx)::Union{Adjoint{Float64,Array{Float64,2}}, Array}
│   %8 = ChainRulesCore.unthunk(%7)::Union{Adjoint{Float64,Array{Float64,2}}, Array}
└──      return %8
```
it looks like there's also an issue in that it's not able to realise that the `Zero` path won't get hit...

Reply:
@oxinabox what are we missing here? This feels like something basic...

Reply:
Interestingly, the above example doesn't seem to be an issue on Julia nightly.